src/nvm.rs

```diff
@@ -307,26 +328,40 @@ impl<T, const N: usize> Collection<T, N> where T: Copy {
         self.capacity() - self.len()
     }
 
-    /// Returns the index of an item in the internal storage, given the index
-    /// in the collection. If index is too big, None is returned.
+    /// Counts the number of allocated slots up until `len`.
+    fn count_allocated(&self, len: usize) -> usize {
+        self.flags.get_ref()
+            .iter()
+            .take(len)
+            .fold(0, |acc, &byte| acc + (byte == STORAGE_VALID) as u32) as usize
+    }
+
+    /// Returns the `key` of an item in the internal storage, given the `index`
+    /// in the collection. If `index` is too big, None is returned.
     ///
     /// # Arguments
     ///
     /// * `index` - Index in the collection
     fn index_to_key(&self, index: usize) -> Option<usize> {
-        let mut next = 0;
-        let mut count = 0;
+        // Neat optimization: start by setting `next` to index,
+        // because we know we could not have found `index` allocated slots beforehand.
+        let mut next_key = index;
+        // Now count the number of allocated slots we have found up
+        // until this `index` (without including the slot at `index` itself).
+        let mut count = self.count_allocated(index);
```

Comment:
This is the result of https://github.com/LedgerHQ/ledger-nanos-sdk/pull/9/commits/c8a1c122481856751aff18809b3fe9684a3bd5f2 and https://github.com/LedgerHQ/ledger-nanos-sdk/pull/9/commits/54f36733e0c094218835ca70add131f1ff0a6c4a . We could, of course, revert those comits and keep the code a little less confusing.
The performance incrase on this one is linear, so I think it's a _nice to have_